File: LeetCode/Max Consecutive Ones III.cpp
/* 
QUESTION:
Given a binary array nums and an integer k, return the maximum number of consecutive 1's in the array 
if you can flip at most k 0's.
*/


class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int n = nums.size(), ans = 0;
        int left = 0, right = 0, zeros = 0;

        // Iterate over nums with a dynamic window
        while(right < n){
            if(nums[right] == 0){
                zeros++;
            }

            // Shrink the window until the current window contains not more than k zeros
            while (left <= right && zeros > k){
                if(nums[left] == 0){
                    zeros--;
                }

                left++;
            }

            // Choose the best answer
            ans = max(ans, right - left + 1);
            right++;
        } 

        return ans;
    }
};

File: LeetCode/1046-max-consecutive-ones-iii/max-consecutive-ones-iii.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        n = len(nums)
        result = 0
        zeros = 0
        left = 0

        for right in range(n):
            if nums[right] == 0:
                zeros += 1
                if zeros > k:
                    while nums[left] != 0:
                        left += 1

                    left += 1
                    zeros -= 1
    
            result = max(result, right - left + 1)
        
        return result
        